ssl/extensions.cc

```diff
@@ -3089,6 +3089,46 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
   return true;
 }
 
+// Adds pha_ext to the client hello
+static bool ext_pha_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
+                                    CBB *out_compressible, ssl_client_hello_type_t type)  {
+  const SSL *const ssl = hs->ssl;
+
+  // post_handshake_auth is not supported in TLS versions < 1.3. Also checking
+  // if extension has been enabled by client.
+  if(hs->min_version < TLS1_3_VERSION || ssl->s3->pha_enabled != 1) {
+    return true;
+  }
+
+  if (!CBB_add_u16(out_compressible, TLSEXT_TYPE_post_handshake_auth) ||
+      !CBB_add_u16(out_compressible, 0 /* empty "extension_data" field */) ||
+      !CBB_flush(out_compressible)) {
+    return false;
+  }
+
+  ssl->s3->pha_ext = SSL_PHA_EXT_SENT;
+  return true;
+}
+
+// Parses pha_ext from client hello
+static bool ext_pha_parse_clienthello(SSL_HANDSHAKE *hs,
+                                      uint8_t *out_alert,
+                                      CBS *contents)  {
+  if(contents == nullptr || hs->min_version < TLS1_3_VERSION) {
+    return true;
+  }
+
+  // Should be empty
+  if(CBS_len(contents) != 0) {
+    return false;
+  }
```

Comment:
If false, the calling context returns an error. False just indicates parsing had an error. So in the above code snippet, if the body of the extension is not of length 0 (which it should be since we constructed it that way in ext_pha_add_clienthello), we'd return false and the calling context (https://github.com/smittal3/aws-lc/blob/4b5c619d7a326ecd673ae6afac2774ad8fc0bb97/ssl/extensions.cc#L3661) would return an error. Otherwise true indicates successful parsing.